### PR TITLE
[MAINTENANCE] Numpy 2 compatibility (initial PR)

### PR DESCRIPTION
**Pull request by anthonyburdi**
Add ruff rule for NPY201, first part of preparing to enable numpy 2.0 compatibility & make resulting changes. 
Changes tested to be compatible with earlier versions of numpy. Our [min supported version is 1.20.3 for python 3.8](https://github.com/great-expectations/great_expectations/blob/ca9346ab9fdd50149951c5d559e9998a0946dc90/ci/constraints-test/py38-min-install.txt#L1). The changes in this PR are supported back to at least that version:
- inf: https://numpy.org/doc/1.20/reference/constants.html?highlight=inf#numpy.inf
- nan: https://numpy.org/doc/1.20/reference/constants.html?highlight=nan#numpy.nan
- double: https://numpy.org/doc/1.20/reference/arrays.scalars.html?highlight=double#numpy.double

Needs https://github.com/great-expectations/great_expectations/pull/9273 to merge first.

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -258,7 +258,12 @@ select = [
     "PL", # pylint
     # https://beta.ruff.rs/docs/rules/#ruff-specific-rules-ruf
     "RUF", # custom ruff rules
+    # Enable after fixing or ignoring errors:
+    # "NPY", # numpy 2.0 compatibility (see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin)
+    "NPY201", # numpy 2.0 compatibility (see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin) -- Explicitly listed to view preview rule errors
 ]
+preview = true
+explicit-preview-rules = true
```
Comment (Kilo59) [inline pyproject.toml]:
```suggestion
explicit-preview-rules = true  # https://docs.astral.sh/ruff/preview/#selecting-single-preview-rules
```

File: contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_follow_rule.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ class ColumnRuleFollowers(ColumnMapMetricProvider):
 
     @column_condition_partial(engine=PandasExecutionEngine)
     def _pandas(cls, column, rule, **kwargs):
-        if rule["ranges"] is {}:
+        if rule["ranges"] == {}:
```
Comment (billdirks) [inline contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_follow_rule.py]: This looks like a stray change. It looks like this does fix a bug. You might want to change this to just `if rule["ranges"]:` or this change is fine too.

Comment (anthonyburdi) [reply contrib/experimental/great_expectations_experimental/expectations/expect_column_values_to_follow_rule.py]: This was due to a ruff rule using the preview version of the rule - see https://github.com/great-expectations/great_expectations/pull/9273 for the fix (removed from this PR).

File: pyproject.toml [hunk 3]
```diff
@@ -258,6 +258,9 @@ select = [
     "PL", # pylint
     # https://beta.ruff.rs/docs/rules/#ruff-specific-rules-ruf
     "RUF", # custom ruff rules
+    # Enable after fixing or ignoring errors:
+    # "NPY", # numpy 2.0 compatibility (see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin)
+    "NPY201", # numpy 2.0 compatibility (see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin) -- Explicitly listed to view preview rule errors
```
Comment (Kilo59) [inline pyproject.toml]:
```suggestion
    #  https://docs.astral.sh/ruff/rules/#numpy-specific-rules-npy
    # "NPY", # NumPy-specific rules TODO: Enable after fixing or ignoring errors
    "NPY201", # numpy 2.0 compatibility (see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#ruff-plugin) -- Explicitly listed to view preview rule errors
```
minor formatting changes to match the other entries in this section

**Review by Kilo59 — APPROVED** [review]
Minor formatting suggestion

**Pull request merged**